solver/directory.go

```diff
@@ -0,0 +1,123 @@
+package solver
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"path/filepath"
+
+	"github.com/docker/buildx/util/progress"
+	"github.com/moby/buildkit/client"
+	"github.com/moby/buildkit/client/llb"
+	gateway "github.com/moby/buildkit/frontend/gateway/client"
+	digest "github.com/opencontainers/go-digest"
+	"github.com/openllb/hlb/parser"
+	"github.com/openllb/hlb/parser/ast"
+	"github.com/openllb/hlb/pkg/llbutil"
+	"golang.org/x/sync/errgroup"
+)
+
+// NewRemoteDirectory returns an ast.Directory representing a directory backed
+// by a BuildKit gateway reference. Since gateway references only live while
+// the build session is running, the build only exits when the ast.Directory is
+// closed.
+func NewRemoteDirectory(ctx context.Context, cln *client.Client, pw progress.Writer, def *llb.Definition, root string, dgst digest.Digest, solveOpts []SolveOption, sessionOpts []llbutil.SessionOption) (ast.Directory, error) {
+	// Block constructing remoteDirectory until the graph is solved and assigned to
+	// ref.
+	resolved := make(chan struct{})
+
+	s, err := llbutil.NewSession(ctx, sessionOpts...)
+	if err != nil {
+		return nil, err
+	}
+
+	g, ctx := errgroup.WithContext(ctx)
+
+	// ctx.Done keeps Build from exiting until remoteDirectory is closed.
+	// This ensures that cache keys and results from the build are not garbage
+	// collected while its still in use.
+	ctx, cancel := context.WithCancel(ctx)
+
+	g.Go(func() error {
+		return s.Run(ctx, cln.Dialer())
+	})
+
+	var ref gateway.Reference
+	g.Go(func() error {
+		return Build(ctx, cln, s, pw, func(ctx context.Context, c gateway.Client) (*gateway.Result, error) {
+			dir, err := c.Solve(ctx, gateway.SolveRequest{
+				Definition: def.ToPB(),
+			})
+			if err != nil {
+				return nil, err
+			}
+
+			ref, err = dir.SingleRef()
+			if err != nil {
+				return nil, err
+			}
+
+			close(resolved)
+			<-ctx.Done()
+			return gateway.NewResult(), nil
+		}, solveOpts...)
+	})
+
+	select {
+	case <-ctx.Done():
+		cancel()
+		return nil, g.Wait()
+	case <-resolved:
+		// If ref is nil, then an error has occurred when solving, clean up and
+		// return.
+		if ref == nil {
```

Comment:
Not sure how this can be reached. If `SingleRef` returns an error, nothing closes `resolved`.